sgx/enclave/src/attestation.rs

```diff
@@ -1,9 +1,167 @@
+    let mut report_data: sgx_report_data_t = sgx_report_data_t::default();
+
+    // Fill ecc256 public key into report_data
+    let mut pub_k_gx = public_key.gx.clone();
+    pub_k_gx.reverse();
+    let mut pub_k_gy = public_key.gy.clone();
+    pub_k_gy.reverse();
+    report_data.d[..32].clone_from_slice(&pub_k_gx);
+    report_data.d[32..].clone_from_slice(&pub_k_gy);
+
+    report_data
+}
+
+fn keypair() -> Result<(sgx_ec256_private_t, sgx_ec256_public_t), sgx_status_t> {
+    let ecc_handle = SgxEccHandle::new();
+    let _result = ecc_handle.open();
+    ecc_handle.create_key_pair()
+}
+
+fn quote_nonce() -> io::Result<sgx_quote_nonce_t> {
+    let mut quote_nonce = sgx_quote_nonce_t { rand : [0;16] };
+    let mut os_rng = SgxRng::new()?;
+    os_rng.fill_bytes(&mut quote_nonce.rand);
+    Ok(quote_nonce)
+}
+
+const RET_QUOTE_BUF_LEN : u32 = 2048;
+type QuoteBuf = [u8; RET_QUOTE_BUF_LEN as usize];
+
+fn quote(report: &sgx_report_t, quote_nonce: &sgx_quote_nonce_t)
+    -> Result<(sgx_report_t, QuoteBuf, u32), sgx_status_t> {
+    let mut qe_report = sgx_report_t::default();
+    let mut quote_buf = [0; RET_QUOTE_BUF_LEN as usize];
+    let mut quote_len : u32 = 0;
+
+    let spid = sgx_spid_t::default();
+    let mut rt = sgx_status_t::SGX_ERROR_UNEXPECTED;
+
+    let result = unsafe {
+        ocall_get_quote(
+            &mut rt as *mut sgx_status_t,
+            ptr::null(), // Not using a signature revocation list yet
+            0,
+            report as * const sgx_report_t,
+            sgx_quote_sign_type_t::SGX_LINKABLE_SIGNATURE,
+            &spid as *const sgx_spid_t,
+            quote_nonce as * const sgx_quote_nonce_t,
+            &mut qe_report as *mut sgx_report_t,
+            quote_buf.as_mut_ptr(),
+            RET_QUOTE_BUF_LEN as u32,
+            &mut quote_len as *mut u32)
+    };
+
+    if result != sgx_status_t::SGX_SUCCESS {
+        println!("ocall_get_quote failed {}", result);
+        return Err(result);
+    }
+
+    if rt != sgx_status_t::SGX_SUCCESS {
+        println!("ocall_get_quote returned {}", rt);
+        return Err(rt);
+    }
+
+    Ok((qe_report, quote_buf, quote_len))
+}
+
+fn report_matches_current_platform(target_info: &sgx_target_info_t, qe_report: &sgx_report_t) -> bool {
+    target_info.mr_enclave.m == qe_report.body.mr_enclave.m &&
+        target_info.attributes.flags == qe_report.body.attributes.flags &&
+        target_info.attributes.xfrm  == qe_report.body.attributes.xfrm
+}
+
+fn report_tampered(qe_report: &sgx_report_t, quote_buf: &QuoteBuf, quote_len: u32, quote_nonce: &sgx_quote_nonce_t) -> bool {
+    let mut rhs_vec : Vec<u8> = quote_nonce.rand.to_vec();
+    rhs_vec.extend(&quote_buf[..quote_len as usize]);
+    let rhs_hash = tcrypto::rsgx_sha256_slice(&rhs_vec[..]).unwrap();
+    let lhs_hash = &qe_report.body.report_data.d[..32];
+
+    rhs_hash != lhs_hash
+}
 
 pub fn report() -> Result<sgx_report_t, sgx_status_t> {
```

Comment:
This seems like the top level method, the main if you will. Are we able to put it at the top of the file for a more top down reading flow?